crypto/bn/bn_shift.c

```diff
@@ -82,40 +83,63 @@ int BN_rshift1(BIGNUM *r, const BIGNUM *a)
 
 int BN_lshift(BIGNUM *r, const BIGNUM *a, int n)
 {
-    int i, nw, lb, rb;
-    BN_ULONG *t, *f;
-    BN_ULONG l;
-
-    bn_check_top(r);
-    bn_check_top(a);
+    int ret;
 
     if (n < 0) {
         BNerr(BN_F_BN_LSHIFT, BN_R_INVALID_SHIFT);
         return 0;
     }
 
+    ret = bn_lshift_fixed_top(r, a, n);
+
+    bn_correct_top(r);
+    bn_check_top(r);
+
+    return ret;
+}
+
+int bn_lshift_fixed_top(BIGNUM *r, const BIGNUM *a, int n)
+{
+    int i, nw, lb, rb;
+    BN_ULONG *t, *f;
+    BN_ULONG l, m, rmask = 0;
+
+    assert(n >= 0);
+
+    bn_check_top(r);
+    bn_check_top(a);
+
     nw = n / BN_BITS2;
     if (bn_wexpand(r, a->top + nw + 1) == NULL)
```

Comment:
Implies that `n` is public? Perhaps that should be mentioned in a leading comment to this function.

Reply:
One has to keep in mind that "fixed-top" is not automatically equal to "constant-time". Caller is as responsible to play the cards right. In this case note that constant-time-ness claim is being made about division, not shifting. So that, yes, `n` is public, and when this subroutine is called from bn_div_fixed_top, it's always zero. But sure, I can think of commentary...

Reply:
Correction! `n / BN_BITS2` is public and is always zero when called from bn_div_fixed_top. `n % BN_BITS2` is private and protected.